walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.lang.String;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Request is too long or too short(more than 1 day/negative duration)
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration() || request.getDuration() <= 0)
+    {
+      return Arrays.asList();
+    }
+
+    // Return full day for trivial empty cases (no event given, no attendees given)
+    if (events.isEmpty() || request.getAttendees().isEmpty())
+    {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    events = sortEventsByStartTime(events);
+
+    int requiredDuration = (int)request.getDuration(); // Can cast as such because duration does not exceed 2^32.
+    Collection<String> requiredAttendees = request.getAttendees();
+    
+    int currentTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    
+    ArrayList<TimeRange> candidateRanges = new ArrayList<TimeRange>();
+
+    Iterator eventsIterator = events.iterator();
+    Event currentEvent = (Event)eventsIterator.next();
+    Event previousEvent = currentEvent;
+    while (startTime != TimeRange.END_OF_DAY) {       
+      
+      boolean areAttendeesInCurrentEvent = !Collections.disjoint(currentEvent.getAttendees(), request.getAttendees());
+
+      if (areAttendeesInCurrentEvent && currentEvent.getWhen().start() >= startTime ) // Available time is only up to when event starts.
+      { 
+        currentTime = currentEvent.getWhen().start();
+
+        TimeRange potentialSlot = TimeRange.fromStartEnd(startTime, currentTime, false);
+        if (potentialSlot.duration() >= requiredDuration)
+        {
+          candidateRanges.add(potentialSlot);
+        }
+
+        startTime = currentEvent.getWhen().end();
+      }       
+      
+      if (eventsIterator.hasNext()) // Current event can be considered as available time. lookahead to next event.
+      {
+        currentEvent = (Event)eventsIterator.next();
+        if (previousEvent.getWhen().overlaps(currentEvent.getWhen()) && areAttendeesInCurrentEvent){ // Next event is intersecting and our start time needs to be set properly.
+          
+          // If the new event end is still during the previous event, we can set the new start time to the previous event's end, as that is a stricter bound.
+          // If the new event end is after the previous event, then we should consider the new event as per normal and start our timings from the start of the new event.
+          startTime = currentEvent.getWhen().end() < previousEvent.getWhen().end() ? previousEvent.getWhen().end() : currentEvent.getWhen().start();
+        }
```

Comment:
It probably is more readable to move this check to the beginning of your while loop. Also, you can simplify this as follows:

`startTime = Math.max(currentEvent.getWhen().end(), previousEvent.getWhen().end())`

You're already checking `areAttendeesInCurrentEvent` so that would mean you can't consider the duration of the currentEvent as a possibility.

Reply:
done.